New version 2.4.0 breaks some code

## Problem with v2.4.0

Given code like this:
```scala
.flatMap { implicit logger: Logger[IO] => 
  program
}
```

When I run scalafmt it gets formatted as below:

```scala
.flatMap(implicit logger: Logger[IO] => program)
```

Which doesn't compile.

## CI Build error

Here's the PR updating scalafmt: https://github.com/profunktor/redis4cats/pull/224

Here's the full CI build error: https://github.com/profunktor/redis4cats/pull/224/checks?check_run_id=447840196

```
[error] /home/runner/work/redis4cats/redis4cats/modules/examples/src/main/scala/dev/profunktor/redis4cats/LoggerIOApp.scala:38:54: '=>' expected but ')' found.
[error]       .flatMap(implicit logger: Logger[IO] => program)
[error]                                                      ^
[error] /home/runner/work/redis4cats/redis4cats/modules/examples/src/main/scala/dev/profunktor/redis4cats/LoggerIOApp.scala:41:1: illegal start of simple expression
[error] }
[error] ^
[error] two errors found
```

I'm facing a similar issue rewriting `{}` to ()`.  On [ths http4s commit](https://github.com/http4s/http4s/tree/d568e34db6441af38eb1ae8763ab9e8faddc940e), after running `core/scalafmtAll`, we get the following diff on `QueryParam.scala`:

```diff
    }
 
   def instantQueryParamEncoder(formatter: DateTimeFormatter): QueryParamEncoder[Instant] =
-    QueryParamEncoder[String].contramap[Instant] { i: Instant =>
-      formatter.format(i)
-    }
+    QueryParamEncoder[String].contramap[Instant](i: Instant => formatter.format(i))
 
   @deprecated("Use QueryParamEncoder[U].contramap(f)", "0.16")
   def encodeBy[T, U](f: T => U)(
```

This fails to compile:

```
sbt:http4s> test:compile
[info] Compiling 32 Scala sources to /home/ross/src/http4s/core/target/scala-2.13/classes ...
[error] /home/ross/src/http4s/core/src/main/scala/org/http4s/QueryParam.scala:94:80: ')' expected but '(' found.
[error]     QueryParamEncoder[String].contramap[Instant](i: Instant => formatter.format(i))
[error]                                                                                ^
[error] /home/ross/src/http4s/core/src/main/scala/org/http4s/QueryParam.scala:94:83: ';' expected but ')' found.
[error]     QueryParamEncoder[String].contramap[Instant](i: Instant => formatter.format(i))
[error]                                                                                   ^
[error] two errors found
/home/ross/src/http4s/core/src/main/scala/org/http4s/QueryParam.scala:94: error: ) expected but ( found
    QueryParamEncoder[String].contramap[Instant](i: Instant => formatter.format(i))
                                                                               ^
[error] (core / Compile / compileIncremental) Compilation failed
[error] Total time: 2 s, completed Feb 15, 2020 11:26:00 AM
```

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala
@@ -174,8 +174,9 @@ class FormatWriter(formatOps: FormatOps) {
             case b: Term.Block if TreeOps.getBlockSingleStat(b).exists {
                   /* guard for statements requiring a wrapper block
                    * "foo { x => y; z }" can't become "foo(x => y; z)" */
-                  case Term.Function(_, body) =>
-                    TreeOps.getTermSingleStat(body).isDefined
+                  case f: Term.Function =>
+                    TreeOps.getTermSingleStat(f.body).isDefined &&
+                      !RedundantBraces.needParensAroundParams(f)
                   case _ => true
                 } =>
               b.parent match {

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala
@@ -13,6 +13,18 @@ import org.scalafmt.util.TreeOps._
 object RedundantBraces extends Rewrite {
   override def create(implicit ctx: RewriteCtx): RewriteSession =
     new RedundantBraces
+
+  def needParensAroundParams(f: Term.Function): Boolean =
+    /* either we have parens or no type; multiple params or
+     * no params guarantee parens, so we look for type and
+     * parens only for a single param */
+    f.params match {
+      case List(param) if param.decltpe.nonEmpty =>
+        val leftParen = f.tokens.find(_.is[Token.LeftParen])
+        !leftParen.exists(_.start <= param.tokens.head.start)
+      case _ => false
+    }
+
 }
 
 /**
@@ -193,13 +205,14 @@ class RedundantBraces(implicit ctx: RewriteCtx) extends RewriteSession {
     }
   }
 
-  private def okToRemoveBlockWithinApply(b: Term.Block): Boolean = {
+  private def okToRemoveBlockWithinApply(b: Term.Block): Boolean =
     getSingleStatIfLineSpanOk(b).exists {
-      case Term.Function(_, _: Term.Block) =>
-        !settings.methodBodies // else the inner block will be rewritten, too
+      case f: Term.Function =>
+        // don't rewrite block if the inner block will be rewritten, too
+        !(f.body.is[Term.Block] && settings.methodBodies) &&
+          !RedundantBraces.needParensAroundParams(f)
       case _ => true
     }
-  }
 
   /** Some blocks look redundant but aren't */
   private def shouldRemoveSingleStatBlock(b: Term.Block): Boolean =

--- a/scalafmt-tests/src/test/resources/rewrite/RedundantBraces-ParenLambdas.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/RedundantBraces-ParenLambdas.stat
@@ -136,20 +136,18 @@ override def run(args: List[String]): IO[ExitCode] =
       program
     }
 >>>
-test does not parse
 override def run(args: List[String]): IO[ExitCode] =
   Slf4jLogger
     .create[IO]
-    .flatMap(implicit logger: Logger[IO] => program)
+    .flatMap { implicit logger: Logger[IO] => program }
 <<< #1707 2: don't rewrite to parens if typed lambda param
 def a(b: B): C[D] =
     C[String].contramap[D] { i: D =>
       b.format(i)
     }
 >>>
-test does not parse
-def a(formatter: DateTimeFormatter): b[Instant] =
-  C[String].contramap[D](i: D => b.format(i))
+def a(b: B): C[D] =
+  C[String].contramap[D] { i: D => b.format(i) }
 <<< #1707 3: rewrite to parens if typed lambda param with parens
 def a(b: B): C[D] =
     C[String].contramap[D] { (i: D) =>
@@ -172,6 +170,4 @@ danglingParentheses = false
 val a = b[IO](
 { c: Int => if (d) e else f }, g)
 >>>
-test does not parse
-val a = b[IO](
- c: Int => if (d) e else f , g)
+val a = b[IO]({ c: Int => if (d) e else f }, g)

--- a/scalafmt-tests/src/test/resources/rewrite/RedundantBraces-ParenLambdas.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/RedundantBraces-ParenLambdas.stat
@@ -128,3 +128,50 @@ object a {
 object a {
   val a = b(c => d => { e; f })
 }
+<<< #1707 1: don't rewrite to parens if typed lambda param
+override def run(args: List[String]): IO[ExitCode] =
+  Slf4jLogger
+    .create[IO]
+    .flatMap { implicit logger: Logger[IO] =>
+      program
+    }
+>>>
+test does not parse
+override def run(args: List[String]): IO[ExitCode] =
+  Slf4jLogger
+    .create[IO]
+    .flatMap(implicit logger: Logger[IO] => program)
+<<< #1707 2: don't rewrite to parens if typed lambda param
+def a(b: B): C[D] =
+    C[String].contramap[D] { i: D =>
+      b.format(i)
+    }
+>>>
+test does not parse
+def a(formatter: DateTimeFormatter): b[Instant] =
+  C[String].contramap[D](i: D => b.format(i))
+<<< #1707 3: rewrite to parens if typed lambda param with parens
+def a(b: B): C[D] =
+    C[String].contramap[D] { (i: D) =>
+      b.format(i)
+    }
+>>>
+def a(b: B): C[D] =
+  C[String].contramap[D]((i: D) => b.format(i))
+<<< #1707 4: rewrite to parens if multiple params
+def a(b: B): C[D] =
+    C[String].contramap[D] { (i: D, j: Int) =>
+      b.format(i)
+    }
+>>>
+def a(b: B): C[D] =
+  C[String].contramap[D]((i: D, j: Int) => b.format(i))
+<<< #1708
+danglingParentheses = false
+===
+val a = b[IO](
+{ c: Int => if (d) e else f }, g)
+>>>
+test does not parse
+val a = b[IO](
+ c: Int => if (d) e else f , g)

--- a/scalafmt-tests/src/test/scala/org/scalafmt/util/CanRunTests.scala
+++ b/scalafmt-tests/src/test/scala/org/scalafmt/util/CanRunTests.scala
@@ -27,7 +27,7 @@ trait CanRunTests extends AnyFunSuite with HasTests {
             } catch {
               case e: ParseException =>
                 fail(
-                  "test does not parse" +
+                  "test does not parse\n" +
                     parseException2Message(e, t.original)
                 )
             }

